rclcpp/src/rclcpp/executor.cpp

```diff
@@ -366,24 +366,36 @@ Executor::spin_some_impl(std::chrono::nanoseconds max_duration, bool exhaustive)
   }
   RCPPUTILS_SCOPE_EXIT(this->spinning.store(false); );
 
-  while (rclcpp::ok(context_) && spinning.load() && max_duration_not_elapsed()) {
-    if (!wait_result_.has_value()) {
-      wait_for_work(std::chrono::milliseconds(0));
-    }
+  // clear the wait result and wait for work without blocking to collect the work
+  // for the first time
+  // both spin_some and spin_all wait for work at the beginning
+  wait_result_.reset();
+  wait_for_work(std::chrono::milliseconds(0));
+  bool just_waited = true;
 
+  while (rclcpp::ok(context_) && spinning.load() && max_duration_not_elapsed()) {
     AnyExecutable any_exec;
     if (get_next_ready_executable(any_exec)) {
       execute_any_executable(any_exec);
+      just_waited = false;
     } else {
-      // If nothing is ready, reset the result to signal we are
-      // ready to wait again
+      // if nothing is ready, reset the result to clear it
       wait_result_.reset();
-    }
 
-    if (!wait_result_.has_value() && !exhaustive) {
-      // In the case of spin some, then we can exit
-      // In the case of spin all, then we will allow ourselves to wait again.
-      break;
+      if (just_waited) {
+        // there was no work after just waiting, always exit in this case
+        // before the exhaustive condition can be checked
+        break;
+      }
+
+      if (exhaustive) {
+        // if exhaustive, wait for work again
+        // this only happens for spin_all; spin_some only waits at the start
+        wait_for_work(std::chrono::milliseconds(0));
+        just_waited = true;
+      } else {
+        break;
+      }
```

Comment:
This is the important part of the change, there was one issue with the logic before:

- when doing spin_all, exhaustive = true, so the break would never occur, even when there was no work to be done
  - this caused spin_all to start busy-waiting once there was no work to be done, the original reported issue

The original patch suggested logic that essentially replaced the `&&` with `||`, but that caused spin_some to not execute all of the data.

So this new logic is correct and hopefully more clear. It does this:

- while not canceled or timeout
- try to get next token of work
- if available execute
- if not, clear wait result and then ...
- if no work was done since the last wait, then break
- else if the exhaustive option is true, wait for more work
- else stop

Reply:
I think it would be useful to capture this in a comment above the implementation, just to make it clear that this sequence should be maintained going forward.  It's very easy to modify code in here without understanding the full impact.

Reply:
see [8587653](https://github.com/ros2/rclcpp/pull/2517/commits/8587653dc4feb6adce1b4eb0c8c1cb79d687ff90)